test/retryOnEmpty.test.ts

```diff
@@ -0,0 +1,668 @@
+              const request: JsonRpcRequest<string[]> = {
+                id: 1,
+                jsonrpc: '2.0',
+                method,
+                params: buildMockParamsWithBlockParamAt(
+                  blockParamIndex,
+                  blockParam,
+                ),
+              };
+              const sendAsyncSpy = stubProviderRequests(provider, [
+                stubBlockNumberRequest(),
+              ]);
+
+              await engine.handle(request);
+
+              expectProviderRequestNotToHaveBeenMade(sendAsyncSpy, request);
+            });
+          },
+        );
+
+        it('does not make a request directly through the provider, but skips straight to the next middleware, if no block parameter is given', async () => {
+          await withTestSetup(async ({ engine, provider }) => {
+            const request: JsonRpcRequest<string[]> = {
+              id: 1,
+              jsonrpc: '2.0',
+              method,
+              params: buildMockParamsWithoutBlockParamAt(blockParamIndex),
+            };
+            const sendAsyncSpy = stubProviderRequests(provider, [
+              stubBlockNumberRequest(),
+            ]);
+
+            await engine.handle(request);
+
+            expectProviderRequestNotToHaveBeenMade(sendAsyncSpy, request);
+          });
+        });
+      });
+    });
+  });
+
+  describe('a method that does not take a block parameter', () => {
+    it('does not make a request directly through the provider, but skips straight to the next middleware', async () => {
+      await withTestSetup(async ({ engine, provider }) => {
+        const method = 'a_non_block_param_method';
+        const request: JsonRpcRequest<string[]> = {
+          id: 1,
+          jsonrpc: '2.0',
+          method,
+        };
+        const sendAsyncSpy = stubProviderRequests(provider, [
+          stubBlockNumberRequest(),
+        ]);
+
+        await engine.handle(request);
+
+        expectProviderRequestNotToHaveBeenMade(sendAsyncSpy, request);
+      });
+    });
+  });
+});
+
+/**
+ * When using `stubProviderRequests` to list canned responses for specific
+ * requests that are made to `provider.sendAsync`, you don't need to provide the
+ * full request object to go along with the response, but only part of that
+ * request object. When `provider.sendAsync` is then called, we can look up the
+ * compare the real request object to the request object that was specified to
+ * find a match. This function is used to do that comparison (and other
+ * like comparisons).
+ *
+ * @param requestMatcher - A partial request object.
+ * @param request - A real request object.
+ * @returns True or false depending on whether the partial request object "fits
+ * inside" the real request object.
+ */
+function requestMatches(
+  requestMatcher: Partial<JsonRpcRequest<unknown>>,
+  request: JsonRpcRequest<unknown>,
+): boolean {
+  return (
+    Object.keys(requestMatcher) as unknown as (keyof typeof requestMatcher)[]
```

Comment:
Nit: I don't think the `as unknown` part here is required:
```suggestion
    Object.keys(requestMatcher)  as (keyof typeof requestMatcher)[]
```

Reply:
Hmm, weird. Not sure why I thought this was necessary. Removed in [e4f373b](https://github.com/MetaMask/eth-json-rpc-middleware/pull/147/commits/e4f373b0462f4b653714ba58c358574a85048f61).